In [842]:
# Nama   : Mia Amaliah 
# NIM    : 1202922002
# Tugas : RFM Analysis

import numpy as np 
import pandas as pd


import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))



/kaggle/input/rfm-uas-data/UAS_RFM.csv


# a. Bentuklah RFM (Recency Frequency Monetery) model untuk segmentasi pengunjung Giant 

In [843]:
df= pd.read_csv('../input/rfm-uas-data/UAS_RFM.csv')
df.head (10)

CardID    TrxDate    Amount
0  1000152  6/14/2021  686481.0
1  1000158  2/16/2021  923251.0
2  1000156  4/22/2021  576485.0
3  1000115  4/12/2021  373726.0
4  1000164  4/30/2021  431905.0
5  1000136  6/29/2021  216754.0
6  1000147   6/5/2021  580528.0
7  1000118  1/12/2021  806750.0
8  1000167   2/8/2021  546173.0
9  1000160  3/21/2021  597906.0

In [844]:
df.shape

(199, 3)

In [845]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199 entries, 0 to 198
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CardID   199 non-null    int64  
 1   TrxDate  199 non-null    object 
 2   Amount   199 non-null    float64
dtypes: float64(1), int64(1), object(1)
memory usage: 4.8+ KB


In [846]:
df.isna().sum()

CardID     0
TrxDate    0
Amount     0
dtype: int64

In [847]:
df['TrxDate'] = pd.to_datetime(df['TrxDate'])

In [848]:
max_date = max(df['TrxDate'])
max_date

Timestamp('2021-06-29 00:00:00')

In [849]:
df['Diff'] = max_date - df['TrxDate']
df.head()

CardID    TrxDate    Amount     Diff
0  1000152 2021-06-14  686481.0  15 days
1  1000158 2021-02-16  923251.0 133 days
2  1000156 2021-04-22  576485.0  68 days
3  1000115 2021-04-12  373726.0  78 days
4  1000164 2021-04-30  431905.0  60 days

In [850]:
type_counts = df['CardID'].value_counts()
CardID=pd.DataFrame(type_counts)
CardID.head()

CardID
1000152       7
1000143       7
1000112       7
1000158       6
1000157       6

In [851]:
RFM = df.groupby('CardID')['Diff'].min().reset_index()
RFM.head()

CardID    Diff
0  1000101 48 days
1  1000102 98 days
2  1000103 16 days
3  1000104 40 days
4  1000105  4 days

In [852]:
RFM['Diff'] = RFM['Diff'].dt.days
RFM.columns = ['CardID','Recency']
RFM.head()

CardID  Recency
0  1000101       48
1  1000102       98
2  1000103       16
3  1000104       40
4  1000105        4

In [853]:
RFM_F = df.groupby('CardID')['TrxDate'].count().reset_index()
RFM_F.columns = ['CardID', 'Frequency']
RFM_F.head()



CardID  Frequency
0  1000101          1
1  1000102          3
2  1000103          1
3  1000104          2
4  1000105          5

In [854]:
RFM_F.tail(5)

CardID  Frequency
66  1000169          1
67  1000170          2
68  1000171          1
69  1000172          3
70  1000174          1

In [855]:
RFM_M = df.groupby('CardID')['Amount'].sum().reset_index()
RFM_M.columns = ['CardID','Amount']
RFM_M.head()

CardID     Amount
0  1000101   194215.0
1  1000102  1185855.0
2  1000103   845228.0
3  1000104   696678.0
4  1000105  2360357.0

In [856]:
RFM_M.tail()

CardID     Amount
66  1000169   484173.0
67  1000170   978678.0
68  1000171   341512.0
69  1000172  1668941.0
70  1000174   494254.0

In [857]:
RFM = RFM.merge(RFM_F,how='inner',on=['CardID'])
RFM =RFM.merge(RFM_M,how='inner',on=['CardID'])
RFM.head(25)

CardID  Recency  Frequency     Amount
0   1000101       48          1   194215.0
1   1000102       98          3  1185855.0
2   1000103       16          1   845228.0
3   1000104       40          2   696678.0
4   1000105        4          5  2360357.0
5   1000106       27          2   792655.0
6   1000107       25          3  1555026.0
7   1000108       33          3  1310430.0
8   1000109       83          2   499011.0
9   1000110       31          5  2764479.0
10  1000111       82          2   745911.0
11  1000112       72          7  4463425.0
12  1000113       49          4  2123945.0
13  1000114       16          4  2411422.0
14  1000115       78          4  2186153.0
15  1000116       49          2   476445.0
16  1000117       76          1   377155.0
17  1000118       51          3  1810549.0
18  1000119       90          2  1448926.0
19  1000120       31          2  1173664.0
20  1000121       80          1   607586.0
21  1000122       17          3  1370572.0
22  1000124      127          3  1351677.0
23  1000125       34          4  2512856.0
24  1000126      152          2   433738.0

In [858]:
RFM.tail(25)

CardID  Recency  Frequency     Amount
46  1000148       52          2   947113.0
47  1000149       34          2  1596495.0
48  1000150        1          3  1860607.0
49  1000151      102          2  1136883.0
50  1000152       15          7  3844381.0
51  1000153      123          1   714119.0
52  1000154        4          6  2435542.0
53  1000155       28          3  2156921.0
54  1000156       68          2  1403417.0
55  1000157       17          6  3065378.0
56  1000158       27          6  3306820.0
57  1000159        6          3  1478997.0
58  1000160       33          2  1533995.0
59  1000161       25          5  3735016.0
60  1000162      139          1   526739.0
61  1000163       35          3  1660175.0
62  1000164       39          2   892682.0
63  1000166       59          6  3817607.0
64  1000167        8          3  1092850.0
65  1000168        9          2   304367.0
66  1000169      111          1   484173.0
67  1000170       78          2   978678.0
68  1000171       59          1   341512.0
69  1000172       23          3  1668941.0
70  1000174      178          1   494254.0

# b.  Tentukan rekomendasi campaign bagi masing-masing segment

Dari segmentasi RFM adalah metode analisis yang membagi segmentasi pelanggan ke dalam 4 kelas, yaitu Gold, Silver, Bronze dan non profit. Penjelasannya :  
1. segmentasi Gold memiliki kualifikasi konsumen dengan monetary tinggi, frequency nya tinggi , dan recency nya juga tinggi
==> jika dilihat dari data untuk userID  dibawah ini bisa termasuk ke dalam golongan segmentasi gold karena jika dilihat dari frekuensi, monetary dan recencynya yang paling tinggi dibandingkan dengan yg lainnya
# 1000152	15	7	3844381.0
# 1000105	4	5	2360357.0

2. segmentasi Silver memiliki kualifikasi konsumen dengan tingkat monetary yang tinggi, frequency yang rendah, dan recency yang tinggi.
==> jika dilihat dari hasil datanya untuk UserID dibawah ini termasuk ke dalam segmentasi silver
# 1000151	102	2	1136883.0

3. segmentasi  Bronze terdiri dari konsumen dengan tingkat monetary rendah, frequency rendah, dan recency yang tinggi. 
==> salah satu userID yang termasuk ke dalam segment ini adalah dengan userid 1000124

# 1000124	127	3	1351677.0
4. Sementara Non-Profit memiliki kualifikasi konsumen dengan kualifikasi monetary, frequency, dan recency yang rendah.
==> jika di lihat dari datanya untuk user id id bawah ini termasuk ke dalam segment non profit

# 1000174	178	1	494254.0